### PR TITLE
Assistive-playwright

### PR DESCRIPTION
**Pull request by divdavem**

### CODE REVIEW

File: components/assistive-playwright-client/src/screenReaderClient.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    ) => T | undefined,
+    options: {
+      includeCurrent?: boolean;
+      timeout?: number;
+    } = {}
+  ): Promise<T> {
+    const { includeCurrent = true, timeout = 10000 } = options;
+    return new Promise<T>((resolve, reject) => {
+      if (includeCurrent) {
+        for (let msg = this.messages, i = 0, l = msg.length; i < l; i++) {
```
Comment (fbasso) [inline components/assistive-playwright-client/src/screenReaderClient.ts:190]: It's a matter of taste (and coding preferences), but for this kind of loop (so here and in several places in this PR), I would have preferred  the `for...of` or `for...in` usage which is quite concise:

for... in here as your need the index in the body:
```
const messages = this.messages;
for(const i in messages) {
   const result = condition(messages[i], i, messages);
   ...
}
```

File: components/assistive-playwright-client/src/screenReaderClient.ts [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+          if (result) {
+            resolve(result);
+            return;
+          }
+        }
+      }
+      const removeListeners = () => {
+        clearTimeout(plannedTimeout);
+        this.off("message", listener);
+        this.on("error", errorListener);
```
Comment (fbasso) [inline components/assistive-playwright-client/src/screenReaderClient.ts:201]: Not sure if it's a typo or not here (there is a lot of code in this PR !), but is it not supposed to be : this.**off**("error", errorListener); ?

Comment (divdavem) [reply components/assistive-playwright-client/src/screenReaderClient.ts:201]: Yes, thank you! I will fix it!

File: components/assistive-playwright-client/src/screenReaderClient.ts (hunk 1 above)
Comment (divdavem) [reply components/assistive-playwright-client/src/screenReaderClient.ts:190]: I could have used `for ... of` if I did not need the index.
But `for ... in` is for the properties of any object, and there is no guarrantee about the order of iteration: cf https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in

Comment (fbasso) [reply components/assistive-playwright-client/src/screenReaderClient.ts:190]: Right, I didn't remember about the order. So it's better to let it like this.

Just a note: the order of a for...in is in the spec of the ES2020, so the browsers should end to respect it (it's probably already the case, but not guarantee).

Comment (divdavem) [reply components/assistive-playwright-client/src/screenReaderClient.ts:190]: Note that this code is not supposed to run in a browser, but in node.js, which, indeed, probably respects the order defined by the spec of ES2020, but I prefer having a clear code for which it is not necessary to wonder if the order will be correct.

**Review by fbasso — APPROVED** [review]
Thanks for your PR @divdavem !

This is a quite big PR, so it's not easy to understand every point of the code. I focused my review mostly on the api proposed to control the mouse and the keyboard, compared on the one given by Playwright, and I'm quite fine with it. 

I think the low level mouse/keyboard interactions coupled with the playwright waiting mechanisms and the waitForMessage you provide is quite good for the stability (action -> wait -> check -> action -> wait -> ... is the common loop to render the tests stable).

So I'm fine with this PR. I was not able to be able to see if there is issues with the code (I tried my best :) ), but if there is, it will be time to open smaller  PRs to fix them !

LGTM for a merge !

**Comment by divdavem** [comment]
@fbasso Thank you very much for your taking the time to review the code! And thank you for your feedback, it is really appreciated!

**Pull request merged**